package/MDAnalysis/core/Selection.py

```diff
@@ -356,181 +336,176 @@ def __init__(self, x, y, z, cutoff):
-    def __init__(self, type):
-        StringSelection.__init__(self, "type")
-        self.type = type
+    field = 'types'
 
 
 class ResidueNameSelection(StringSelection):
-    def __init__(self, resname):
-        StringSelection.__init__(self, "resname")
-        self.resname = resname
+    field = 'resnames'
 
 
 class SegmentNameSelection(StringSelection):
-    def __init__(self, segid):
-        StringSelection.__init__(self, "segid")
-        self.segid = segid
+    field = 'segids'
 
 
 class AltlocSelection(StringSelection):
-    def __init__(self, altLoc):
-        StringSelection.__init__(self, "altLoc")
-        self.altLoc = altLoc
+    field = 'altLocs'
 
 
-class ByResSelection(Selection):
+class ByResSelection(object):
     def __init__(self, sel):
         self.sel = sel
 
-    def _apply(self, group):
-        res = self.sel._apply(group)
-        unique_res = set([(a.resid, a.segid) for a in res])
-        sel = []
-        for atom in group.atoms:
-            if (atom.resid, atom.segid) in unique_res:
-                sel.append(atom)
-        return set(sel)
+    def apply(self, group):
+        logger.debug("In ByResSelection")
+        res = self.sel.apply(group)
+        logger.debug("Reference group is {}".format(res))
+
+        unique_res = np.unique(res.resids)
+        logger.debug("Ref resindices are {}".format(unique_res))
+        mask = np.in1d(group.resids, unique_res)
 
+        return unique(group[mask])
 
-class _RangeSelection(Selection):
+
+class RangeSelection(object):
     def __init__(self, lower, upper):
         self.lower = lower
         self.upper = upper
 
 
-class ResidueIDSelection(_RangeSelection):
-    def _apply(self, group):
+class ResidueIDSelection(RangeSelection):
+    def apply(self, group):
+        resids = group.resids
         if self.upper is not None:
-            return set([a for a in group.atoms if (self.lower <= a.resid <= self.upper)])
+            mask = resids >= self.lower
+            mask &= resids <= self.upper
         else:
-            return set([a for a in group.atoms if a.resid == self.lower])
+            mask = resids == self.lower
+        return unique(group[mask])
 
 
-class ResnumSelection(_RangeSelection):
-    def _apply(self, group):
+class ResnumSelection(RangeSelection):
+    def apply(self, group):
+        resnums = group.resnums
         if self.upper is not None:
-            return set([a for a in group.atoms if (self.lower <= a.resnum <= self.upper)])
+            mask = resnums >= self.lower
+            mask &= resnums <= self.upper
         else:
-            return set([a for a in group.atoms if a.resnum == self.lower])
+            mask = resnums == self.lower
+        return unique(group[mask])
 
 
-class ByNumSelection(_RangeSelection):
-    def _apply(self, group):
+class ByNumSelection(RangeSelection):
+    def apply(self, group):
+        # In this case we'll use 1 indexing since that's what the
```

Comment:
Does this fix a bug? (I.e. something for CHANGELOG)

Can you make sure that the docs are clear on what "bynum" does?